File: DataStructures/trees/PreInPostOrderBST.java
import java.util.*;
import java.io.*;

class Node{
	Node left;
	Node right;
	int data;

	Node(int data){
		this.data = data;
		this.left = null;
		this.right = null;
	}
}

class PreInPostOrderBST{

	public static void preOrder(Node root){
		if(root != null){
            System.out.print(root.data+ " ");
            if(root.left != null){
                preOrder(root.left);
            }
            if(root.right != null){
                preOrder(root.right);
            }
        }
	}

	public static void postOrder(Node root) {
        if(root != null){
            if(root.left != null){
                postOrder(root.left);
            }
            if(root.right !=null){
            postOrder(root.right);
            }
            System.out.print(root.data+ " ");
        }
    }

    public static void inOrder(Node root) {
        if(root != null){
            if(root.left !=null){
                inOrder(root.left);
            }
            System.out.print(root.data+" ");
            if(root.right != null){
                inOrder(root.right);
            }
        }
    }

	public static Node insert(Node root,int data){
		if(root == null) {
            return new Node(data);
        } else {
            Node cur;
            if(data <= root.data) {
                cur = insert(root.left, data);
                root.left = cur;
            } else {
                cur = insert(root.right, data);
                root.right = cur;
            }
            return root;
        }
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
        int t = scan.nextInt();
        Node root = null;
        while(t-- > 0) {
            int data = scan.nextInt();
            root = insert(root, data);
        }
        scan.close();
        System.out.print("preOrder: ");
        preOrder(root);
        System.out.print("\npostorder: ");
        postOrder(root);
        System.out.print("\ninOrder: ");
        inOrder(root);
	}
}

File: DataStructures/trees/preinpostorderBST.py
#o(log n) or o(n)
class Node:
	def __init__(self, info):
		self.info = info
		self.left = None
		self.right = None
		self.level = None

	def __str__(self):
		return str(self.info)

class BinarySearchTree:
	def __init__(self):
		self.root = None

	def create(self, val):
		if self.root == None:
			self.root = Node(val)
		else:
			current = self.root
			while True:
				if val < current.info:
					if current.left:
						current = current.left
					else:
						current.left = Node(val)
						break
				elif val > current.info:
					if current.right:
						current = current.right
					else:
						current.right = Node(val)
						break
				else:
					break

def preOrder(root):
	if root:
		print(root.info,end=" ")
		if root.left:
			preOrder(root.left)
		if root.right:
			preOrder(root.right)

def inOrder(root):
	if root:
		if root.left:
			inOrder(root.left)
		print(root.info,end=" ")
		if root.right:
			inOrder(root.right)

def postOrder(root):
	if root:
		if root.left:
			postOrder(root.left)
		if root.right:
			postOrder(root.right)
		print(root.info,end=" ")



tree = BinarySearchTree()
t = int(input())

arr = list(map(int,input().split()))

for i in range(t):
	tree.create(arr[i])

print("preOrder: ",end="")
preOrder(tree.root)
print()
print("inOrder: ",end="")
inOrder(tree.root)
print()
print("postOrder: ",end="")
postOrder(tree.root)
